markets/storageadapter/dealpublisher.go

```diff
@@ -443,3 +467,83 @@ func (p *DealPublisher) filterCancelledDeals() {
 	}
 	p.pending = filtered
 }
+
+func (p *DealPublisher) queueDeals(pdeal *pendingDeal) {
+	p.lk.Lock()
+	defer p.lk.Unlock()
+
+	// Filter out any cancelled deals
+	p.filterCancelledDeals()
+
+	// Make sure the new deal hasn't been cancelled
+	if pdeal.ctx.Err() != nil {
+		return
+	}
+
+	// Add the new deal to the queue
+	p.pending = append(p.pending, pdeal)
+	log.Infof("add deal with piece CID %s to publish deals queue - %d deals in queue",
+		pdeal.deal.Proposal.PieceCID, len(p.pending))
+}
+
+func (p *DealPublisher) PublishQueuedDeals(deals []cid.Cid) (error, []cid.Cid) {
```

Comment:
I'd suggest we combine PublishQueuedDeals with the existing publishAllDeals method. Instead of calling publishAddDeals, the existing code should call PublishQueuedDeals but pass all deals to it.